File: src/main/java/team404/handlers/InventoryListener.java
package team404.handlers;

import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.format.TextColor;
import net.kyori.adventure.text.serializer.plain.PlainTextComponentSerializer;
import net.kyori.adventure.title.Title;
import org.apache.commons.lang3.tuple.Pair;
import org.bukkit.Bukkit;
import org.bukkit.EntityEffect;
import org.bukkit.GameMode;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.OfflinePlayer;
import org.bukkit.Particle;
import org.bukkit.Sound;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.block.Action;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.event.player.PlayerInteractEvent;
import org.bukkit.event.player.PlayerJoinEvent;
import org.bukkit.inventory.Inventory;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.SkullMeta;
import org.bukkit.plugin.Plugin;
import org.bukkit.scheduler.BukkitRunnable;
import team404.PlayerRevivalService;
import team404.models.requests.PlayerRevivedRequest;
import team404.utils.TextUtils;
import team404.constants.ColorHexConstants;
import team404.models.Recipes;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;

import static team404.constants.InventoryConstants.INVENTORY_NAME;
import static team404.constants.MessagesConstants.NOT_ENOUGH_RESOURCES;
import static team404.constants.MessagesConstants.PLAYER_ALREADY_REVIVED;
import static team404.constants.MessagesConstants.PLAYER_REVIVED;
import static team404.constants.MessagesConstants.YOU_WILL_BE_REVIVED_IN;
import static team404.constants.TelegramBotConstants.PLAYER_REVIVED_PATH;
import static team404.utils.HttpClient.sendPostRequest;

public class InventoryListener implements Listener {
    private static final String WORLD_NAME = "world";
    private static final int SECONDS_BEFORE_RESPAWN = 5;
    private static final int INVENTORY_ROW_SIZE = 9;
    private final Plugin plugin;
    private final PlayerRevivalService playerRevivalService;

    public InventoryListener(Plugin plugin) {
        this.plugin = plugin;
        this.playerRevivalService = PlayerRevivalService.getInstance(plugin);
    }

    @EventHandler
    public void onPlayerJoin(PlayerJoinEvent event) {
        OfflinePlayer player = event.getPlayer();
        if (playerRevivalService.getPlayersToSpawn().contains(player.getName())) {
            respawnPlayer(player);
        }
    }

    @EventHandler
    public void onRightClick(PlayerInteractEvent event) {
        if (event.getAction() != Action.RIGHT_CLICK_AIR && event.getAction() != Action.RIGHT_CLICK_BLOCK) {
            return;
        }
        ItemStack item = event.getItem();
        if (!Recipes.isReviveStuff(item)) {
            return;
        }
        Player player = event.getPlayer();
        Inventory inventory = getInventory(player);
        player.openInventory(inventory);
    }

    private Inventory getInventory(Player player) {
        Inventory inv = Bukkit.createInventory(null, getInventorySize(), Component.text(INVENTORY_NAME));
        for (Map.Entry<String, List<Pair<Integer, Material>>> entry : playerRevivalService.getRespawnablePlayers().entrySet()) {
            ItemStack playerHead = new ItemStack(Material.PLAYER_HEAD);
            SkullMeta skullMeta = (SkullMeta) playerHead.getItemMeta();
            skullMeta.setOwningPlayer(Bukkit.getOfflinePlayer(entry.getKey()));
            List<Component> lore = new ArrayList<>();
            for (Pair<Integer, Material> pair : entry.getValue()) {
                ItemStack name = new ItemStack(pair.getRight());
                if (checkMaterialInInventory(player, pair)) {
                    lore.add(TextUtils.appendCheckMark(name)
                            .appendSpace()
                            .append(Component.text(pair.getLeft())));
                } else {
                    lore.add(TextUtils.appendCross(name)
                            .appendSpace()
                            .append(Component.text(pair.getLeft())));
                }
            }
            skullMeta.lore(lore);
            playerHead.setItemMeta(skullMeta);
            inv.addItem(playerHead);
        }
        return inv;
    }

    private int getInventorySize() {
        var players = playerRevivalService.getRespawnablePlayers();
        if (players.isEmpty()) {
            return INVENTORY_ROW_SIZE;
        }
        int rowsCount = players.size() / INVENTORY_ROW_SIZE;
        if (players.size() % INVENTORY_ROW_SIZE != 0) {
            rowsCount++;
        }
        return rowsCount * INVENTORY_ROW_SIZE;
    }

    private boolean checkMaterialInInventory(Player player, Pair<Integer, Material> material) {
        int requiredAmount = material.getLeft();
        Material requiredMaterial = material.getRight();

        int count = 0;
        for (ItemStack item : player.getInventory().getContents()) {
            if (item != null && item.getType() == requiredMaterial) {
                count += item.getAmount();
            }
        }

        return count >= requiredAmount;
    }

    public boolean checkMaterialInInventory(Player player, List<Pair<Integer, Material>> materials) {
        if (materials == null) {
            return false;
        }
        for (Pair<Integer, Material> material : materials) {
            if (!checkMaterialInInventory(player, material)) {
                return false;
            }
        }
        return true;
    }

    @EventHandler
    public void onInventoryClick(InventoryClickEvent event) {
        if (!(event.getWhoClicked() instanceof Player player) ||
                !PlainTextComponentSerializer.plainText().serialize(event.getView().title()).equals(INVENTORY_NAME)) {
            return;
        }
        ItemStack clickedItem = event.getCurrentItem();
        if (clickedItem == null) {
            return;
        }
        event.setCancelled(true); // Prevents taking items from the inventory
        // check that head clicked
        if (clickedItem.getType() != Material.PLAYER_HEAD) {
            return;
        }

        OfflinePlayer offlinePlayerToSpawn = ((SkullMeta) clickedItem.getItemMeta()).getOwningPlayer();
        // check that player not revived yet
        var playerToSpawnName = offlinePlayerToSpawn.getName();
        if (!playerRevivalService.getRespawnablePlayers().containsKey(playerToSpawnName)) {
            player.sendMessage(PLAYER_ALREADY_REVIVED);
            return;
        }
        // check that player who clicked has enough materials
        List<Pair<Integer, Material>> materials = playerRevivalService.getRespawnablePlayers().get(playerToSpawnName);
        if (!checkMaterialInInventory(player, materials)) {
            player.sendMessage(Component.text(NOT_ENOUGH_RESOURCES)
                    .color(TextColor.fromHexString(ColorHexConstants.RED_HEX)));
            return;
        }
        // add player to spawn map
        playerRevivalService.addPlayerToRespawn(playerToSpawnName);
        playerRevivalService.removeRespawnablePlayer(playerToSpawnName);

        player.sendMessage(
                Component.text(PLAYER_REVIVED.formatted(playerToSpawnName))
                        .color(TextColor.fromHexString(ColorHexConstants.GREEN_HEX))
        );
        player.closeInventory();
        removeItems(player, materials);
        // try respawn
        respawnPlayer(offlinePlayerToSpawn);
    }

    private void respawnPlayer(OfflinePlayer offlinePlayer) {
        final AtomicInteger counter = new AtomicInteger(SECONDS_BEFORE_RESPAWN);
        if (!offlinePlayer.isOnline()) {
            return;
        }
        Player player = offlinePlayer.getPlayer();
        new BukkitRunnable() {
            @Override
            public void run() {
                int currentCounter = counter.getAndDecrement();
                player.clearTitle();
                if (currentCounter > 0) {
                    player.showTitle(
                            Title.title(
                                    Component.text(YOU_WILL_BE_REVIVED_IN.formatted(currentCounter)),
                                    Component.empty()
                            )
                    );
                } else {
                    // check that player is online
                    if (offlinePlayer.isOnline()) {
                        // if online teleport and set gamemode and remove from to spawn list
                        Location location = offlinePlayer.getRespawnLocation();
                        location = location != null ? location : Bukkit.getWorld(WORLD_NAME).getSpawnLocation();
                        player.teleport(location);
                        player.setGameMode(GameMode.SURVIVAL);
                        player.spawnParticle(Particle.TOTEM_OF_UNDYING, location, 100);
                        player.playEffect(EntityEffect.TOTEM_RESURRECT);
                        player.playSound(location, Sound.ENTITY_PLAYER_LEVELUP, 5, 1);
                        player.clearActivePotionEffects();
                        playerRevivalService.removePlayerToRespawn(player.getName());
                        sendPostRequest(
                                plugin.getConfig().getString("bot-address") + PLAYER_REVIVED_PATH,
                                new PlayerRevivedRequest(player.getName())
                        );
                    }
                    // if not do nothing
                    cancel();
                }
            }
        }.runTaskTimer(plugin, 0, 20);
    }

    private void removeItems(Player player, List<Pair<Integer, Material>> itemsList) {
        for (Pair<Integer, Material> amountMaterialPair : itemsList) {
            removeItems(player, amountMaterialPair);
        }
    }

    private void removeItems(Player player, Pair<Integer, Material> itemToRemove) {
        int remaining = itemToRemove.getLeft();

        for (ItemStack item : player.getInventory().getContents()) {
            if (item != null && item.getType().equals(itemToRemove.getRight())) {
                int inStack = item.getAmount();
                if (inStack > remaining) {
                    item.setAmount(inStack - remaining);
                    return;
                } else {
                    player.getInventory().remove(item);
                    remaining -= inStack;
                }
                if (remaining <= 0) {
                    return;
                }
            }
        }
    }
}

File: src/main/kotlin/team404/listener/InventoryListener.kt
package team404.listener

import net.kyori.adventure.text.Component
import net.kyori.adventure.text.format.TextColor
import net.kyori.adventure.text.serializer.plain.PlainTextComponentSerializer
import org.bukkit.Bukkit
import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.block.Action
import org.bukkit.event.inventory.InventoryClickEvent
import org.bukkit.event.player.PlayerInteractEvent
import org.bukkit.inventory.Inventory
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.meta.SkullMeta
import org.bukkit.plugin.Plugin
import org.slf4j.LoggerFactory
import team404.constant.BUYBACK_INVENTORY_NAME
import team404.constant.CANT_BUYBACK
import team404.constant.GOLDEN_HEX
import team404.constant.GREEN_HEX
import team404.constant.INVENTORY_ROW_SIZE
import team404.constant.MAX_BUYBACK_COUNT
import team404.constant.NOT_ENOUGH_RESOURCES
import team404.constant.PLAYER_ALREADY_REVIVED
import team404.constant.PLAYER_BOUGHT_BACK
import team404.constant.PLAYER_REVIVED
import team404.constant.PLUGIN_NAMESPACE
import team404.constant.RED_HEX
import team404.constant.REVIVE_INVENTORY_NAME
import team404.constant.YOU_ARE_FREE
import team404.model.Recipes
import team404.service.PlayerRevivalService
import team404.util.TextUtils
import team404.util.TextUtils.formatTime
import team404.util.getBuybackCount
import team404.util.getBuybackTimeLeft
import team404.util.getBuybackTimerBar
import team404.util.isInBuyback
import team404.util.removeBuybackCount
import team404.util.removeBuybackTimeLeft
import team404.util.removeItems
import team404.util.setIsInBuyback


class InventoryListener(plugin: Plugin) : Listener {
    private val logger = LoggerFactory.getLogger(PLUGIN_NAMESPACE)

    private val playerRevivalService: PlayerRevivalService = PlayerRevivalService.getInstance(plugin)

    @EventHandler
    fun onRightClick(event: PlayerInteractEvent) {
        if (event.action != Action.RIGHT_CLICK_AIR && event.action != Action.RIGHT_CLICK_BLOCK) {
            return
        }
        val item = event.item
        if (!Recipes.isReviveStuff(item)) {
            return
        }
        event.player.also { player ->
            val inventory = getReviveInventory(player)
            player.openInventory(inventory)
        }
    }

    private fun getReviveInventory(player: Player): Inventory {
        val inv = Bukkit.createInventory(null, getReviveInventorySize(), Component.text(REVIVE_INVENTORY_NAME))
        playerRevivalService.respawnablePlayers.forEach { (name, materials) ->
            val headOwner = Bukkit.getOfflinePlayer(name)
            val playerHead = ItemStack(Material.PLAYER_HEAD)
            val skullMeta = playerHead.itemMeta as SkullMeta
            skullMeta.owningPlayer = headOwner
            val lore = materials.map { (material, amount) ->
                val item = ItemStack(material)
                val text = if (checkMaterialInInventory(player, material, amount)) {
                    TextUtils.appendCheckMark(item)
                } else {
                    TextUtils.appendCross(item)
                }
                text.appendSpace().append(Component.text(amount))
            }.toMutableList()
            headOwner.player?.getBuybackTimeLeft()
                ?.let {
                    lore.add(Component.text("Выкупается").color(TextColor.fromHexString(GOLDEN_HEX)))
                    lore.add(
                        Component.text("Осталось времени: ${formatTime(it)}").color(TextColor.fromHexString(GOLDEN_HEX))
                    )
                }
            skullMeta.lore(lore)
            playerHead.itemMeta = skullMeta
            inv.addItem(playerHead)
        }
        return inv
    }

    private fun getReviveInventorySize(): Int {
        val players = playerRevivalService.respawnablePlayers
        if (players.isEmpty()) {
            return INVENTORY_ROW_SIZE
        }
        val rowsCount = players.size / INVENTORY_ROW_SIZE + if (players.size % INVENTORY_ROW_SIZE != 0) 1 else 0
        return rowsCount * INVENTORY_ROW_SIZE
    }

    private fun checkMaterialInInventory(player: Player, requiredMaterial: Material, requiredAmount: Int): Boolean {
        var count = 0
        for (item in player.inventory.contents) {
            if (item != null && item.type == requiredMaterial) {
                count += item.amount
            }
        }
        return count >= requiredAmount
    }

    private fun checkMaterialInInventory(player: Player, materials: Map<Material, Int>?): Boolean {
        materials ?: return false
        return materials.all { (material, amount) -> checkMaterialInInventory(player, material, amount) }
    }

    @EventHandler
    fun onInventoryClick(event: InventoryClickEvent) {
        val reviver = event.whoClicked as? Player ?: return
        when (PlainTextComponentSerializer.plainText().serialize(event.view.title())) {
            REVIVE_INVENTORY_NAME -> {
                val clickedItem = event.currentItem ?: return
                event.isCancelled = true // Prevents taking items from the inventory
                if (clickedItem.type != Material.PLAYER_HEAD) {
                    return
                }
                val offlinePlayerToSpawn = (clickedItem.itemMeta as SkullMeta).owningPlayer ?: return

                val isPlayerBoughtback = offlinePlayerToSpawn.isInBuyback()
                if (reviver.uniqueId != offlinePlayerToSpawn.uniqueId && isPlayerBoughtback) {
                    reviver.sendMessage(PLAYER_BOUGHT_BACK)
                    return
                }

                val playerToSpawnName = offlinePlayerToSpawn.name ?: return
                if (!playerRevivalService.respawnablePlayers.containsKey(playerToSpawnName) || playerRevivalService.playersToSpawn.contains(playerToSpawnName)) {
                    reviver.sendMessage(PLAYER_ALREADY_REVIVED)
                    return
                }

                val materials = playerRevivalService.respawnablePlayers[playerToSpawnName] ?: return
                if (!checkMaterialInInventory(reviver, materials)) {
                    reviver.sendMessage(Component.text(NOT_ENOUGH_RESOURCES).color(TextColor.fromHexString(RED_HEX)))
                    return
                }

                playerRevivalService.removeRespawnablePlayer(playerToSpawnName)

                reviver.closeInventory()
                removeItems(reviver, materials)

                if (isPlayerBoughtback && reviver.uniqueId == offlinePlayerToSpawn.uniqueId) {
                    reviver.setIsInBuyback(false)
                    reviver.removeBuybackCount()
                    reviver.removeBuybackTimeLeft()
                    reviver.getBuybackTimerBar()?.apply {
                            this.removeAll()
                            Bukkit.removeBossBar(this.key)
                        }
                    reviver.sendMessage(Component.text(YOU_ARE_FREE).color(TextColor.fromHexString(GOLDEN_HEX)))
                } else {
                    reviver.removeBuybackCount()
                    reviver.removeBuybackTimeLeft()
                    reviver.sendMessage(
                        Component.text(PLAYER_REVIVED.format(playerToSpawnName))
                            .color(TextColor.fromHexString(GREEN_HEX))
                    )
                    playerRevivalService.addPlayerToRespawn(playerToSpawnName)
                    playerRevivalService.respawnPlayer(offlinePlayerToSpawn)
                }
            }

            BUYBACK_INVENTORY_NAME -> {
                val clickedItem = event.currentItem ?: return
                event.isCancelled = true
                if (clickedItem.type != Material.PLAYER_HEAD) {
                    return
                }

                val buybackCount = reviver.getBuybackCount()
                if (buybackCount != null && buybackCount >= MAX_BUYBACK_COUNT) {
                    reviver.sendMessage(CANT_BUYBACK)
                    return
                }

                val playerToSpawnName = reviver.name
                if (!playerRevivalService.respawnablePlayers.containsKey(playerToSpawnName)) {
                    reviver.sendMessage(PLAYER_ALREADY_REVIVED)
                    return
                }

                playerRevivalService.addPlayerToRespawn(playerToSpawnName)
                reviver.sendMessage(
                    Component.text(PLAYER_REVIVED.format(playerToSpawnName))
                        .color(TextColor.fromHexString(GREEN_HEX))
                )
                reviver.closeInventory()
                playerRevivalService.buybackPlayer(reviver)
            }
        }
    }
}